File: interviewbit/Hashing/Fraction.cpp
typedef long long int lli;

string solve(lli a, lli b, set<pair<lli,lli>>&hashset, string response){
    if(hashset.find(make_pair(a,b)) != hashset.end()){
        return "(" + response + ")";
    }
        
    if(a == 0)
        return response;
    
    hashset.insert(make_pair(a,b));

    bool gogo = false;
    while(a<b){
        a*=10;
        if(gogo)
            response += "0";
        gogo=true;
    }

    response += to_string(a/b);
    
    return solve(a%b, b, hashset, response);
}

string Solution::fractionToDecimal(int A, int B) {
    lli a =  (lli) A, b = (lli) B;
    string sign = ((a>0 and b<0) or (a<0 and b>0))?"-":"";
    a = abs(a), b = abs(b);
    
    set<pair<lli,lli>>hashset;

    if(a%b == 0)
        return sign+to_string(a/b);
    else {
        return sign+to_string(a/b) + "." + solve(a%b,b, hashset, "");
    }
}

File: interviewbit/Hashing/Fraction.py
class Solution:
    def __init__(self):
        self.hashmap = {}
        self.ans = []
        self.flag = None

    def fractionToDecimal(self,A,B):
        if B == 0:
            return ''
        if A == 0:
            return 0

        response = ""
        if (A>0 and B<0) or (A<0 and B>0):
            response += "-"
        A = abs(A)
        B = abs(B)

        level = 1
        while True:
            if A == 0:
                break
            if A in self.hashmap:
                self.flag = level - self.hashmap[A]
                break
            if level > 1:
                self.hashmap[A] = level
                A *= 10
            self.ans.append(int(A/B))
            A=A%B
            level += 1

        if self.flag is not None:
            response += str(self.ans[0]) + "." + "".join([str(x) for x in self.ans[1:-self.flag]]) + "(" + "".join([str(x) for x in self.ans[-self.flag:]]) + ")"
        else:
            response += str(self.ans[0])
            if len(self.ans) > 1:
                response += "." + "".join([str(x) for x in self.ans[1:]])

        return response
